src/app-layer-detect-proto.c

```diff
@@ -556,7 +556,11 @@ static AppProto AppLayerProtoDetectPPGetProto(Flow *f, const uint8_t *buf, uint3
     } else {
         /* first try the destination port */
         pp_port_dp = AppLayerProtoDetectGetProbingParsers(alpd_ctx.ctx_pp, ipproto, dp);
-        alproto_masks = &f->probing_parser_toclient_alproto_masks;
+        if (dir == idir) {
+            // do not update alproto_masks to let a chance to second packet
+            // for instance when sending a junk packet to a DNS server
+            alproto_masks = &f->probing_parser_toclient_alproto_masks;
```

Comment:
do we have a check that flags when there is a mismatch in the same direction between the 1st and 2nd packet? Will this somehow lead to an event?

Reply:
I am not sure I understand your question.
In one direction, I do not see how there can be any mismatch between 1st and 2nd packet (or whatever the number)

We keep running the probing parsers as long as they return `ALPROTO_UNKNOWN` (aka maybe I recognize this, maybe I do not)
Once a probing parser returned something else, it is not run a second time.

Did I miss something ?

Reply:
Maybe I'm misunderstanding the logic, but the comments speaks of a "second packet" so this implies a first packet that could yield different results?

Reply:
> the comments speaks of a "second packet"

Meaning a packet _in the other direction_, after the first packet in the first direction